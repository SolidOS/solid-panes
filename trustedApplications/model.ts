export type Mode = 'read' | 'append' | 'write' | 'control'

export interface TrustedApplication {
  origin: string
  subject: string
  modes: Mode[]
}
