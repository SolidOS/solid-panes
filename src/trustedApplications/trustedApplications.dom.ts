import { NamedNode, Statement, sym } from 'rdflib'
import { ns, store } from 'solid-ui'
import { generateRandomString, getStatementsToAdd, getStatementsToDelete } from './trustedApplications.utils'

interface FormElements {
  modes: HTMLInputElement[]
  // This appears to be used to store either a node from the store,
  // or a reference to the input (checkbox) element for a particular mode.
  // These typings were created post-hoc, so I'm not sure if that was intentional.
  // Thus, this union type should be considered as descriptive rather than prescriptive.
  origin: undefined | NamedNode | HTMLInputElement
}

export function createApplicationTable (subject: NamedNode) {
  const applicationsTable = createElement('table', {
    class: 'results'
  })

  // creating headers
  const header = createContainer('tr', [
    createText('th', 'Application URL'),
    createText('th', 'Access modes'),
    createText('th', 'Actions')
  ])
  applicationsTable.appendChild(header);

  // creating rows
  (store.each(subject, ns.acl('trustedApp'), undefined, undefined) as Statement[])
    .flatMap(app => store
      .each(app, ns.acl('origin'), undefined, undefined)
      .map(origin => ({
        appModes: store.each(app, ns.acl('mode'), undefined, undefined),
        origin
      })))
    .sort((a: any, b: any) => (a.origin.value < b.origin.value ? -1 : 1))
    .forEach(
      ({ appModes, origin }: { appModes: NamedNode[]; origin: NamedNode }) =>
        applicationsTable.appendChild(
          createApplicationEntry(subject, origin, appModes, updateTable)
        )
    )

  // adding a row for new applications
  applicationsTable.appendChild(
    createApplicationEntry(subject, null, [ns.acl('Read')], updateTable)
  )

  return applicationsTable

  function updateTable () {
    applicationsTable.parentElement!.replaceChild(
      createApplicationTable(subject),
      applicationsTable
    )
  }
}

function createApplicationEntry (
  subject: NamedNode,
  origin: NamedNode | null,
  appModes: NamedNode[],
  updateTable: () => void
): HTMLTableRowElement {
  const trustedApplicationState = {
    origin,
    appModes,
    formElements: {
      modes: [],
      origin: undefined
    } as FormElements
  }
  return createContainer('tr', [
    createContainer('td', [
      createContainer(
        'form',
        [
          createElement(
            'input',
            {
              class: 'textinput',
              placeholder: 'Write new URL here',
              value: origin ? origin.value : ''
            },
            {},
            element => {
              trustedApplicationState.formElements.origin = element
            }
          )
        ],
        {},
        {
          submit: addOrEditApplication
        }
      )
    ]),
    createContainer('td', [
      createContainer(
        'form',
        createModesInput(trustedApplicationState),
        {},
        {
          submit: addOrEditApplication
        }
      )
    ]),
    createContainer('td', [
      createContainer(
        'form',
        origin
          ? [
            createText('button', 'Update', {
              class: 'controlButton',
              style: 'background: LightGreen;'
            }),
            createText(
              'button',
              'Delete',
              {
                class: 'controlButton',
                style: 'background: LightCoral;'
              },
              {
                click: removeApplication
              }
            )
          ]
          : [
            createText('button', 'Add', {
              class: 'controlButton',
              style: 'background: LightGreen;'
            })
          ],
        {},
        {
          submit: addOrEditApplication
        }
      )
    ])
  ])

  function addOrEditApplication (event: Event) {
    event.preventDefault()
    let origin
    try {
      origin = sym(trustedApplicationState.formElements.origin!.value)
    } catch (err) {
      return alert('Please provide an application URL you want to trust')
    }

    const modes = trustedApplicationState.formElements.modes
      .filter(checkbox => checkbox.checked)
      .map(checkbox => checkbox.value)

    const deletions = getStatementsToDelete(
      trustedApplicationState.origin || origin,
      subject,
      store,
      ns
    )
    const additions = getStatementsToAdd(
      origin,
      generateRandomString(),
      modes,
      subject,
      ns
    );
    (store as any).updater.update(deletions, additions, handleUpdateResponse)
  }

  function removeApplication (event: Event) {
    event.preventDefault()
    let origin
    try {
      origin = sym(trustedApplicationState.formElements.origin!.value)
    } catch (err) {
      return alert(
        'Please provide an application URL you want to remove trust from'
      )
    }

    const deletions = getStatementsToDelete(origin, subject, store, ns)
    ;(store as any).updater.update(deletions, [], handleUpdateResponse)
  }

  function handleUpdateResponse (uri: any, success: boolean, errorBody: any) {
    if (success) {
      return updateTable()
    }
    console.error(uri, errorBody)
  }
}

function createElement<K extends keyof HTMLElementTagNameMap> (
  elementName: K,
  attributes: { [name: string]: string } = {},
  eventListeners: { [eventName: string]: EventListener } = {},
  onCreated: null | ((createdElement: HTMLElementTagNameMap[K]) => void) = null
) {
  const element = document.createElement(elementName)
  if (onCreated) {
    onCreated(element)
  }
  Object.keys(attributes).forEach(attName => {
    element.setAttribute(attName, attributes[attName])
  })
  Object.keys(eventListeners).forEach(eventName => {
    element.addEventListener(eventName, eventListeners[eventName])
  })
  return element
}

export function createContainer<K extends keyof HTMLElementTagNameMap> (
  elementName: K,
  children: HTMLElement[],
  attributes = {},
  eventListeners = {},
  onCreated = null
) {
  const element = createElement(
    elementName,
    attributes,
    eventListeners,
    onCreated
  )
  children.forEach(child => element.appendChild(child))
  return element
}

export function createText<K extends keyof HTMLElementTagNameMap> (
  elementName: K,
  textContent: string | null,
  attributes = {},
  eventListeners = {},
  onCreated = null
) {
  const element = createElement(
    elementName,
    attributes,
    eventListeners,
    onCreated
  )
  element.textContent = textContent
  return element
}

function createModesInput ({ appModes, formElements }: {
  appModes: NamedNode[]
  formElements: FormElements
}) {
  return ['Read', 'Write', 'Append', 'Control'].map(mode => {
    const isChecked = appModes.some(
      appMode => appMode.value === ns.acl(mode).value
    )
    return createContainer('label', [
      createElement(
        'input',
        {
          type: 'checkbox',
          ...(isChecked ? { checked: '' } : {}),
          value: ns.acl(mode).uri
        },
        {},
        element => formElements.modes.push(element)
      ),
      createText('span', mode)
    ])
  })
}
