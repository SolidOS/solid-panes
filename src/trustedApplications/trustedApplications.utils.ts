import { BlankNode, IndexedFormula, NamedNode, st, Literal, sym, Statement } from 'rdflib'
import { Namespaces } from 'solid-namespace'

export function getStatementsToDelete (
  origin: NamedNode | Literal,
  person: NamedNode,
  kb: IndexedFormula,
  ns: Namespaces
): any {
  const applicationStatements = kb.statementsMatching(
    null,
    ns.acl('origin'),
    origin
  )
  const statementsToDelete = applicationStatements.reduce(
    (memo, st) => {
      return memo
        .concat(
          kb.statementsMatching(
            person,
            ns.acl('trustedApp'),
            st.subject as NamedNode
          )
        )
        .concat(kb.statementsMatching(st.subject))
    },
    [] as Array<Statement>
  )
  return statementsToDelete
}

export function getStatementsToAdd (
  origin: NamedNode,
  nodeName: string,
  modes: string[],
  person: NamedNode,
  ns: Namespaces
): any {
  const application = new BlankNode(`bn_${nodeName}`)
  return [
    st(person, ns.acl('trustedApp'), application, person.doc()),
    st(application, ns.acl('origin'), origin, person.doc()),
    ...modes
      .map(mode => sym(mode))
      .map(mode => st(application, ns.acl('mode'), mode, person.doc()))
  ]
}

export function generateRandomString () {
  return Math.random().toString(36).substr(2, 5)
}
