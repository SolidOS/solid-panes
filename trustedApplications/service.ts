import $rdf, { NamedNode, IndexedFormula, Statement } from 'rdflib'
import { Namespaces } from 'solid-namespace'
import { Mode, TrustedApplication } from './model'

export function getStatementsToDelete (
  origin: NamedNode,
  person: NamedNode,
  kb: IndexedFormula,
  ns: Namespaces
) {
  const applicationStatements = kb.statementsMatching(null, ns.acl('origin'), origin, null)
  const statementsToDelete = applicationStatements.reduce(
    (memo, st) => {
      return memo
        .concat(kb.statementsMatching(person, ns.acl('trustedApp'), st.subject, null, false))
        .concat(kb.statementsMatching(st.subject, null, null, null, false))
    },
    [] as Statement[]
  )
  return statementsToDelete
}

export function getStatementsToAdd (
  origin: NamedNode,
  nodeName: string,
  modes: Mode[],
  person: NamedNode,
  ns: Namespaces
) {
  var application = new $rdf.BlankNode(`bn_${nodeName}`)
  return [
    $rdf.st(person, ns.acl('trustedApp'), application, person.doc()),
    $rdf.st(application, ns.acl('origin'), origin, person.doc()),
    ...modes
      .map(mode => {
        return ns.acl(mode)
      })
      .map(mode => $rdf.st(application, ns.acl('mode'), mode, person.doc()))
  ]
}

/* istanbul ignore next [This executes the actual HTTP requests, which is too much effort to test.] */
export function fetchTrustedApps (
  store: $rdf.IndexedFormula,
  subject: $rdf.NamedNode,
  ns: Namespaces
): TrustedApplication[] {
  return (store.each(subject, ns.acl('trustedApp'), undefined, undefined) as any)
    .flatMap((app: $rdf.NamedNode) => {
      return store.each(app, ns.acl('origin'), undefined, undefined)
        .map((origin) => {
          const modes = store.each(app, ns.acl('mode'), undefined, undefined)
          const trustedApp: TrustedApplication = {
            origin: origin.value,
            subject: subject.value,
            modes: modes.map((mode) => deserialiseMode(mode as $rdf.NamedNode, ns))
          }
          return trustedApp
        })
    })
    .sort((appA: TrustedApplication, appB: TrustedApplication) => (appA.origin > appB.origin) ? 1 : -1)
}

/**
 * @param serialisedMode The full IRI of a mode
 * @returns A plain text string representing that mode, i.e. 'read', 'append', 'write' or 'control'
 */
export function deserialiseMode (serialisedMode: $rdf.NamedNode, ns: Namespaces): Mode {
  const deserialisedMode = serialisedMode.value
    .replace(ns.acl('read').value, 'read')
    .replace(ns.acl('append').value, 'append')
    .replace(ns.acl('write').value, 'write')
    .replace(ns.acl('control').value, 'control')

  return deserialisedMode as Mode
}
