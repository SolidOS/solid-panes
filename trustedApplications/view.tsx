import * as React from 'react'
import $rdf, { NamedNode } from 'rdflib'
import vocab from 'solid-namespace'
import { TrustedApplication, Mode } from './model'

const ns = vocab($rdf)

type AddOrUpdate = (origin: string, modes: Mode[]) => Promise<void>
interface Props {
  apps: Array<Exclude<TrustedApplication, { subject: NamedNode}>>;
  onSaveApp: AddOrUpdate;
  onDeleteApp: (origin: string) => Promise<void>;
}

export const View: React.FC<Props> = (props) => {
  return (
    <>
      <p>Here you can manage the applications you trust.</p>,
      <table className="results">
        <thead>
          <tr>
            <th>Application URL</th>
            <th>Access modes</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {props.apps.map((app) => (
            <ApplicationRow
              key={app.origin}
              app={app}
              onSave={props.onSaveApp}
              onDelete={props.onDeleteApp}
            />
          ))}
        </tbody>
      </table>
      <NewApplication onSave={props.onSaveApp}/>
      <h4>Notes</h4>
      <ol>
        <li>Trusted applications will get access to all resources that you have access to.</li>
        <li>You can limit which modes they have by default.</li>
        <li>They will not gain more access than you have.</li>
      </ol>
      <p>
        Application URLs must be valid URL.
        Examples are http://localhost:3000, https://trusted.app, and https://sub.trusted.app.
      </p>
    </>
  )
}

const ApplicationRow: React.FC<{
  app: TrustedApplication,
  onSave: AddOrUpdate,
  onDelete: (origin: string) => Promise<void>
}> = (props) => {
  const initialModes = {
    read: props.app.modes.indexOf('read') !== -1,
    append: props.app.modes.indexOf('append') !== -1,
    write: props.app.modes.indexOf('write') !== -1,
    control: props.app.modes.indexOf('control') !== -1
  }
  const [modes, setModes] = React.useState<{[ key: string]: boolean}>(initialModes)

  const setMode = (mode: Mode, checked: boolean) => {
    setModes({ ...modes, [mode]: checked })
  }
  const getCheckboxHandler = (mode: Mode) => {
    return (event: React.ChangeEvent<HTMLInputElement>) => setMode(mode, event.target.checked)
  }

  const submitHandler = (event: React.FormEvent) => {
    event.preventDefault()

    const newModes = Object.keys(modes).filter(mode => modes[mode]) as Mode[]
    props.onSave(props.app.origin, newModes)
  }

  return (
    <tr>
      <td>
        <p>{props.app.origin}</p>
      </td>
      <td>
        <form onSubmit={submitHandler}>
          <div className="input-wrap">
            <label className="checkbox">
              <input
                type="checkbox"
                onChange={getCheckboxHandler('read')}
                checked={modes.read}
                name="read"
                id="read"
              />
              Read
            </label>
            <label className="checkbox">
              <input
                type="checkbox"
                onChange={getCheckboxHandler('append')}
                checked={modes.append}
                name="append"
                id="append"
              />
              Append
            </label>
            <label className="checkbox">
              <input
                type="checkbox"
                onChange={getCheckboxHandler('write')}
                checked={modes.write}
                name="write"
                id="write"
              />
              Write
            </label>
            <label className="checkbox">
              <input
                type="checkbox"
                onChange={getCheckboxHandler('control')}
                checked={modes.control}
                name="control"
                id="control"
              />
              Control
            </label>
          </div>
          <button type="submit">Save</button>
        </form>
      </td>
      <td>
        <button onClick={() => props.onDelete(props.app.origin)}>Delete</button>
      </td>
    </tr>
  )
}

const NewApplication: React.FC<{ onSave: AddOrUpdate }> = (props) => {
  const [origin, setOrigin] = React.useState()
  const [modes, setModes] = React.useState<{[ key: string]: boolean}>({
    read: false,
    append: false,
    write: false,
    control: false
  })

  const setMode = (mode: Mode, checked: boolean) => {
    setModes({ ...modes, [mode]: checked })
  }
  const getCheckboxHandler = (mode: Mode) => {
    return (event: React.ChangeEvent<HTMLInputElement>) => setMode(mode, event.target.checked)
  }

  const submitHandler = (event: React.FormEvent) => {
    event.preventDefault()

    const newModes = Object.keys(modes).filter(mode => modes[mode]) as Mode[]
    props.onSave(origin, newModes)
  }

  return (
    <form onSubmit={submitHandler}>
      <div className="input-wrap">
        <label>App URL:</label>
        <input
          type="url"
          onChange={(e) => setOrigin(e.target.value)}
          name="origin"
          id="origin"
          placeholder="https://example.com"
        />
      </div>
      <div className="input-wrap">
        <label className="checkbox">
          <input
            type="checkbox"
            onChange={getCheckboxHandler('read')}
            name="read"
            id="read"
          />
          Read
        </label>
        <label className="checkbox">
          <input
            type="checkbox"
            onChange={getCheckboxHandler('append')}
            name="append"
            id="append"
          />
          Append
        </label>
        <label className="checkbox">
          <input
            type="checkbox"
            onChange={getCheckboxHandler('write')}
            name="write"
            id="write"
          />
          Write
        </label>
        <label className="checkbox">
          <input
            type="checkbox"
            onChange={getCheckboxHandler('control')}
            name="control"
            id="control"
          />
          Control
        </label>
      </div>
      <button type="submit">Add</button>
    </form>
  )
}
