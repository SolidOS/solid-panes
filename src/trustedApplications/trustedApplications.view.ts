import { authn, icons, store, widgets } from 'solid-ui'
import { NamedNode } from 'rdflib'

import { PaneDefinition } from 'pane-registry'
import { createApplicationTable, createContainer, createText } from './trustedApplications.dom'

const thisColor = '#418d99'

const trustedApplicationView: PaneDefinition = {
  global: true,
  icon: `${icons.iconBase}noun_15177.svg`,
  name: 'trustedApplications',
  label: () => null,
  render: (subject, context) => {
    const dom = context.dom
    const div = dom.createElement('div')
    div.classList.add('trusted-applications-pane')
    div.setAttribute(
      'style',
      'border: 0.3em solid ' +
      thisColor +
      '; border-radius: 0.5em; padding: 0.7em; margin-top:0.7em;'
    )
    const table = div.appendChild(dom.createElement('table'))
    const main = table.appendChild(dom.createElement('tr'))
    const bottom = table.appendChild(dom.createElement('tr'))
    const statusArea = bottom.appendChild(dom.createElement('div'))
    statusArea.setAttribute('style', 'padding: 0.7em;')
    render(dom, main, statusArea).catch(err => statusArea.appendChild(widgets.errorMessageBlock(dom, err)))
    return div
  }
}

async function render (dom, main, statusArea): Promise<void> {
  const authContext = await authn.logInLoadProfile({ dom: dom, div: main, statusArea: statusArea, me: null })
  const subject = authContext.me as NamedNode

  const profile = subject.doc()
  const editable = store.updater.editable(profile.uri, store)

  main.appendChild(createText('h3', 'Manage your trusted applications'))

  if (!editable) {
    main.appendChild(
      widgets.errorMessageBlock(dom, `Your profile ${subject.doc().uri} is not editable, so we cannot do much here.`)
    )
    return
  }

  main.appendChild(createText('p', 'Here you can manage the applications you trust.'))

  const applicationsTable = createApplicationTable(subject)
  main.appendChild(applicationsTable)

  main.appendChild(createText('h4', 'Notes'))
  main.appendChild(
    createContainer('ol', [
      main.appendChild(
        createText(
          'li',
          'Trusted applications will get access to all resources that you have access to.'
        )
      ),
      main.appendChild(
        createText('li', 'You can limit which modes they have by default.')
      ),
      main.appendChild(
        createText('li', 'They will not gain more access than you have.')
      )
    ])
  )
  main.appendChild(
    createText('p', 'Application URLs must be valid URL. Examples are http://localhost:3000, https://trusted.app, and https://sub.trusted.app.')
  )
}

export default trustedApplicationView
