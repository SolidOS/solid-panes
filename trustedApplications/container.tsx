import * as React from 'react'
import $rdf from 'rdflib'
import vocab from 'solid-namespace'
import { View } from './view'
import { ContainerProps } from '../types'
import { TrustedApplication, Mode } from './model'
import { getStatementsToAdd, getStatementsToDelete, fetchTrustedApps } from './service'

const ns = vocab($rdf)

export const Container: React.FC<ContainerProps> = (props) => {
  if (!props.session) {
    return <div>You are not logged in</div>
  }

  const isEditable: boolean = (props.store as any).updater.editable(props.subject.doc().uri, props.store)
  if (!isEditable) {
    return <div>Your profile {props.subject.doc().uri} is not editable, so we cannot do much here.</div>
  }

  const fetchedTrustedApps: TrustedApplication[] = fetchTrustedApps(props.store, props.subject, ns)

  const [trustedApps, setTrustedApps] = React.useState(fetchedTrustedApps)

  const addOrEditApp = (origin: string, modes: Mode[]) => {
    const result = new Promise<void>((resolve) => {
      const deletions = getStatementsToDelete($rdf.sym(origin), props.subject, props.store, ns)
      const additions = getStatementsToAdd($rdf.sym(origin), generateRandomString(), modes, props.subject, ns)
      props.store.updater!.update(deletions, additions, () => {
        const newApp: TrustedApplication = { subject: props.subject.value, origin, modes }
        setTrustedApps(insertTrustedApp(newApp, trustedApps))
        resolve()
      })
    })

    return result
  }

  const deleteApp = (origin: string) => {
    const result = new Promise<void>((resolve) => {
      const deletions = getStatementsToDelete($rdf.sym(origin), props.subject, props.store, ns)
      props.store.updater!.update(deletions, [], () => {
        setTrustedApps(removeTrustedApp(origin, trustedApps))
        resolve()
      })
    })

    return result
  }

  return (
    <section>
      <View
        apps={trustedApps}
        onSaveApp={addOrEditApp}
        onDeleteApp={deleteApp}
      />
    </section>
  )
}

function insertTrustedApp (app: TrustedApplication, into: TrustedApplication[]): TrustedApplication[] {
  const index = into.findIndex(found => found.origin === app.origin)
  if (index === -1) {
    return into.concat(app)
  }

  return into.slice(0, index)
    .concat(app)
    .concat(into.slice(index + 1))
}
function removeTrustedApp (origin: string, from: TrustedApplication[]): TrustedApplication[] {
  const index = from.findIndex(found => found.origin === origin)
  return (index === -1)
    ? from
    : from.slice(0, index).concat(from.slice(index + 1))
}

function generateRandomString (): string {
  return Math.random().toString(36).substring(7)
}
