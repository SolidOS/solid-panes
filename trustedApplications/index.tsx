/*   Profile Editing Pane
**
** Unlike most panes, this is available any place whatever the real subject,
** and allows the user to edit their own profile.
**
** Usage: paneRegistry.register('profile/profilePane')
** or standalone script adding onto existing mashlib.
*/

import * as React from 'react'
import * as ReactDOM from 'react-dom'
import solidUi, { SolidUi } from 'solid-ui'
import { IndexedFormula } from 'rdflib'
import paneRegistry from 'pane-registry'

import { PaneDefinition } from '../types'
import { Container } from './container'

const nodeMode = (typeof module !== 'undefined')

let panes
let UI: SolidUi

if (nodeMode) {
  UI = solidUi
  panes = paneRegistry
} else { // Add to existing mashlib
  panes = (window as any).panes
  UI = panes.UI
}

const kb: IndexedFormula = UI.store

const thisPane: PaneDefinition = {
  icon: UI.icons.iconBase + 'noun_15177.svg', // Looks like an A - could say it's for Applications?

  name: 'trustedApplications',

  label: function (subject) {
    var types = kb.findTypeURIs(subject)
    if (types[UI.ns.foaf('Person').uri] || types[UI.ns.vcard('Individual').uri]) {
      return 'Manage your trusted applications'
    }
    return null
  },

  render: function (subject, _dom) {
    const container = document.createElement('div')
    UI.authn.solidAuthClient.currentSession().then((session: any) => {
      ReactDOM.render(
        <Container store={UI.store} subject={subject} session={session}/>,
        container
      )
    })

    return container
  }
}

export default thisPane
if (!nodeMode) {
  console.log('*** patching in live pane: ' + thisPane.name)
  panes.register(thisPane)
}
