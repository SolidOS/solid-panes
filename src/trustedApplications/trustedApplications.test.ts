/* eslint-env jest */
import * as $rdf from 'rdflib'
import solidNamespace from 'solid-namespace'
import { generateRandomString, getStatementsToDelete, getStatementsToAdd } from './trustedApplications.utils'

const ns = solidNamespace($rdf)

describe('generateRandomString', () => {
  it('generates a random string five characters long', () => {
    expect(generateRandomString().length).toBe(5)
  })
})

describe('getStatementsToDelete', () => {
  it('should return an empty array when there are no statements', () => {
    const mockStore = $rdf.graph()
    const mockOrigin = $rdf.sym('https://origin.example')
    const mockProfile = $rdf.sym('https://profile.example#me')
    expect(
      getStatementsToDelete(mockOrigin, mockProfile, mockStore, ns)
    ).toEqual([])
  })

  it('should return all statements for the given origin', () => {
    const mockStore = $rdf.graph()
    const mockApplication = $rdf.sym('https://app.example')
    const mockOrigin = $rdf.sym('https://origin.example')
    const mockProfile = $rdf.sym('https://profile.example#me')
    mockStore.add(mockApplication, ns.acl('origin'), mockOrigin)
    mockStore.add(mockApplication, ns.acl('mode'), ns.acl('Read'))
    mockStore.add(mockProfile, ns.acl('trustedApp'), mockApplication)
    const statementsToDelete = getStatementsToDelete(
      mockOrigin,
      mockProfile,
      mockStore,
      ns
    )
    expect(statementsToDelete.length).toBe(3)
    expect(statementsToDelete).toMatchSnapshot()
  })

  it('should not return statements for a different origin', () => {
    const mockStore = $rdf.graph()
    const mockApplication = $rdf.sym('https://app.example')
    const mockOrigin = $rdf.sym('https://origin.example')
    const mockProfile = $rdf.sym('https://profile.example#me')
    mockStore.add(mockApplication, ns.acl('origin'), mockOrigin)
    mockStore.add(mockApplication, ns.acl('mode'), ns.acl('Read'))
    mockStore.add(mockProfile, ns.acl('trustedApp'), mockApplication)

    const statementsToDelete = getStatementsToDelete(
      $rdf.lit('A different origin'), // @@ TODO Remove casting
      mockProfile,
      mockStore,
      ns
    )
    expect(statementsToDelete.length).toBe(0)
    expect(statementsToDelete).toEqual([])
  })
})

describe('getStatementsToAdd', () => {
  it('should return all required statements to add the given permissions for a given origin', () => {
    const mockOrigin = $rdf.sym('https://origin.example')
    const mockProfile = $rdf.sym('https://profile.example#me')
    const modes = [ns.acl('Read').value, ns.acl('Write').value]

    const statementsToAdd = getStatementsToAdd(
      mockOrigin,
      'mock_app_id',
      modes,
      mockProfile,
      ns
    )
    expect(statementsToAdd.length).toBe(4)
    expect(statementsToAdd).toMatchSnapshot()
  })
})
