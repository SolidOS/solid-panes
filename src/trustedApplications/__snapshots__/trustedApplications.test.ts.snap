// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStatementsToAdd should return all required statements to add the given permissions for a given origin 1`] = `
Array [
  Statement {
    "graph": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://profile.example",
    },
    "object": BlankNode {
      "classOrder": 6,
      "isBlank": 1,
      "isVar": 1,
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "predicate": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#trustedApp",
    },
    "subject": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://profile.example#me",
    },
  },
  Statement {
    "graph": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://profile.example",
    },
    "object": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://origin.example",
    },
    "predicate": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#origin",
    },
    "subject": BlankNode {
      "classOrder": 6,
      "isBlank": 1,
      "isVar": 1,
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
  },
  Statement {
    "graph": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://profile.example",
    },
    "object": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#Read",
    },
    "predicate": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": BlankNode {
      "classOrder": 6,
      "isBlank": 1,
      "isVar": 1,
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
  },
  Statement {
    "graph": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://profile.example",
    },
    "object": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#Write",
    },
    "predicate": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": BlankNode {
      "classOrder": 6,
      "isBlank": 1,
      "isVar": 1,
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
  },
]
`;

exports[`getStatementsToDelete should return all statements for the given origin 1`] = `
Array [
  Statement {
    "graph": DefaultGraph {
      "classOrder": undefined,
      "termType": "DefaultGraph",
      "uri": "chrome:theSession",
      "value": "",
    },
    "object": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://app.example",
    },
    "predicate": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#trustedApp",
    },
    "subject": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://profile.example#me",
    },
  },
  Statement {
    "graph": DefaultGraph {
      "classOrder": undefined,
      "termType": "DefaultGraph",
      "uri": "chrome:theSession",
      "value": "",
    },
    "object": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://origin.example",
    },
    "predicate": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#origin",
    },
    "subject": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://app.example",
    },
  },
  Statement {
    "graph": DefaultGraph {
      "classOrder": undefined,
      "termType": "DefaultGraph",
      "uri": "chrome:theSession",
      "value": "",
    },
    "object": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#Read",
    },
    "predicate": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": NamedNode {
      "classOrder": 5,
      "termType": "NamedNode",
      "value": "https://app.example",
    },
  },
]
`;
